keps/sig-node/3815-pod-condition-for-configuration-errors/README.md

```diff
@@ -0,0 +1,1019 @@
+What are some important details that didn't come across above?
+Go in to as much detail as necessary here.
+This might be a good place to talk about core concepts and how they relate.
+-->
+
+### Risks and Mitigations
+
+The impact will be a new condition that gets set to True if a user has a error during startup of their pod.
+<!--
+What are the risks of this proposal, and how do we mitigate? Think broadly.
+For example, consider both security and how this will impact the larger
+Kubernetes ecosystem.
+
+How will security be reviewed, and by whom?
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+
+## Design Details
+
+Add a FailingToStart condition in core/v1:
+
+```go
+const (
+  ...
+	// PodFailingToStart Condition if Pod is stuck due
+	// to configuration error.
+	PodFailingToStart PodConditionType = "FailingToStart"
+)
+```
+
+For errors that have a reason, we can create a condition for PodFailingToStart as true based on the container status.  For any case where reason is set, we can just match reason and generate a PodFailingToStart condition of true.
+
+There are two areas where we set ContainerStatus reason:
+
+[Image-Errors](https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/images/types.go#L27)
+```golang
+	// ErrImageInspect - Unable to inspect image
+	ErrImageInspect = errors.New("ImageInspectError")
+
+	// ErrImagePull - General image pull error
+	ErrImagePull = errors.New("ErrImagePull")
+
+	// ErrImageNeverPull - Required Image is absent on host and PullPolicy is NeverPullImage
+	ErrImageNeverPull = errors.New("ErrImageNeverPull")
+
+	// ErrRegistryUnavailable - Get http error when pulling image from registry
+	ErrRegistryUnavailable = errors.New("RegistryUnavailable")
+
+	// ErrInvalidImageName - Unable to parse the image name.
+	ErrInvalidImageName = errors.New("InvalidImageName")
+```
+- Should we include all these errors as potential failure cases for pending pods?  I can replicate ErrImageNeverPull and InvalidImageName but not sure how to reproduce the other errors.
+
+[kuberuntime_container errors](https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/kuberuntime/kuberuntime_container.go#L59)
+
+```golang
+var (
+	// ErrCreateContainerConfig - failed to create container config
+	ErrCreateContainerConfig = errors.New("CreateContainerConfigError")
+	// ErrPreCreateHook - failed to execute PreCreateHook
+	ErrPreCreateHook = errors.New("PreCreateHookError")
+	// ErrCreateContainer - failed to create container
+	ErrCreateContainer = errors.New("CreateContainerError")
+	// ErrPreStartHook - failed to execute PreStartHook
+	ErrPreStartHook = errors.New("PreStartHookError")
+	// ErrPostStartHook - failed to execute PostStartHook
+	ErrPostStartHook = errors.New("PostStartHookError")
+)
+```
+- If these are set, can we assume that container failed?  Not sure how to recreate failures of states other than ErrCreateContainerConfig
```

Comment:
Another question for a reviewer: Are these internal errors that we should catch for a condition?  Or do they occur often?